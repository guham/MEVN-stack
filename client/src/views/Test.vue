<template>
  <div class="text-center">
    <h1>{{ $t('title') }}</h1>
    <p class="p-2">
      <span>>></span> {{ value }}
    </p>
  </div>
</template>

<i18n>
{
  "en": {
    "title": "Example: fetch a data from the API"
  },
  "fr": {
    "title": "Exemple : récupérer une donnée à partir de l'API"
  },
  "es": {
    "title": "Ejemplo: obtener un dato del API"
  }
}
</i18n>

<script>
import testClient from '@/api/test';
import _get from 'lodash/get';

export default {
  name: 'Test',
  metaInfo: {
    title: 'Test',
  },
  data: () => ({
    value: 'Waiting for value...',
  }),
  mounted() {
    this.fetchValueFromServer();
  },
  methods: {
    fetchValueFromServer() {
      return testClient.fetchValueFromServer()
        .then((response) => {
          this.value = response.data.data;
        }).catch((error) => {
          this.value = _get(error, 'data.message', _get(error, 'message', ''));
        });
    },
  },
};
</script>
