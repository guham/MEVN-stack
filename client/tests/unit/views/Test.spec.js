import axios from 'axios';
import Test from '@/views/Test.vue';
import factory from '../factory';

beforeEach(() => {
  jest.resetModules();
  jest.clearAllMocks();
});

axios.get.mockImplementation(() => Promise.resolve({ data: { data: 'message from the API' } }));

describe('Test.vue', () => {
  test('renders a default message', () => {
    const wrapper = factory(Test);
    expect(wrapper.find('p').text()).toBe('>> Waiting for value...');
  });

  test('has a mounted hook', () => {
    expect(typeof Test.mounted).toBe('function');
  });

  test('once mounted, an API call is done', async () => {
    const wrapper = factory(Test);
    await wrapper.vm.fetchValueFromServer();
    expect(axios.get).toBeCalled();
  });

  test('once mounted, should renders the value fetched from the API', async () => {
    const wrapper = factory(Test);
    await wrapper.vm.fetchValueFromServer();
    expect(wrapper.find('p').text()).toBe('>> message from the API');
  });

  test('API endpoint should be called with "/api/foo/test"', async () => {
    const wrapper = factory(Test);
    await wrapper.vm.fetchValueFromServer();
    expect(axios.get).toHaveBeenCalledWith('/api/foo/test', {});
  });

  test("renders error's message when an error occurs", async () => {
    const wrapper = factory(Test);
    axios.get.mockImplementation(() => Promise.reject(new Error('Error')));
    await wrapper.vm.fetchValueFromServer();
    expect(wrapper.find('p').text()).toBe('>> Error');
  });

  test('has the expected html structure', () => {
    const wrapper = factory(Test);
    const template = wrapper.html();
    expect(template).toMatchSnapshot();
  });
});
