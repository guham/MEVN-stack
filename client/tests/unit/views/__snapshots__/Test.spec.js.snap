// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test.vue has the expected html structure 1`] = `
<div class="text-center">
  <h1>title</h1>
  <p class="p-2"><span>&gt;&gt;</span> Waiting for value...</p>
</div>
`;
